src/wrapper_template.h

```diff
@@ -177,19 +205,166 @@ inline nb::object wrap_multiargument_function(const MultiargumentFunc& func, con
+ * @param input  A vector of nanobind objects, each representing an argument. Each argument can be a scalar,
+ *               list, or ndarray. Lists and ndarrays are expanded; scalars are treated as single values.
+ * @return       A nested nanobind list (nb::object) containing the results of applying func to each
+ *               combination of arguments from the cartesian product.
+ *
+ * Differs from wrap_multiargument_function in that this function computes the cartesian product of argument
+ * lists/arrays, applying the function to every possible combination, whereas wrap_multiargument_function
+ * applies the function to a single set of arguments (possibly vectorized).
+ */
+inline nb::object wrap_cartesian_product_function(const MultiargumentFunc& func, const std::vector<nb::object>& input) {
+  // Parse the input object
+  auto [array_inputs, output_shape] = parse_input(input);
+
+  // Record the size of every input, return empty np.array in case any of the inputs is empty
+  std::vector<size_t> shape;
+
+  for (const auto& input_vector : array_inputs) {
+    ssize_t size = input_vector.size();
+    if (size <= 0) {
+      return nb::ndarray<double, nb::numpy>(nullptr, {0}).cast();
+    }
+    shape.push_back(size);
+  }
+
+  // Initialize the vector of pointers and compute the number of elements in the output
+  int num_inputs = shape.size();
+  std::vector<size_t> index_pointers(num_inputs, 0);
+  int output_size = 1;
+
+  for (size_t i = 0; i < num_inputs; ++i) {
+    output_size *= shape[i];
+  }
+
+  // Iterate through all the combinations and fill the array with functions output
+  double* results = new double[output_size];
+  std::vector<std::variant<double, int>> args(num_inputs);
+
+  auto assign_val = [&](nb::handle val, int j) {
+    if (nb::isinstance<nb::float_>(val)) {
+      args[j] = nb::cast<double>(val);
+    } else if (nb::isinstance<nb::int_>(val)) {
+      args[j] = nb::cast<int>(val);
+    } else {
+      throw nb::type_error("All arguments must be int or float at the deepest level.");
+    }
+  };
+
+  for (int j = 0; j < num_inputs; ++j) {
+    auto val = array_inputs[j][index_pointers[j]];
+    assign_val(val, j);
+  }
+
+  results[0] = func(args);
+
+  for (int i = 1; i < output_size; ++i) {
+    int j = num_inputs - 1;
+
+    while (index_pointers[j] >= shape[j] - 1) {
+      index_pointers[j] = 0;
+      auto val = array_inputs[j][index_pointers[j]];
+      assign_val(val, j);
+      --j;
+    }
+
+    index_pointers[j] += 1;
+    auto val = array_inputs[j][index_pointers[j]];
+    assign_val(val, j);
+
+    results[i] = func(args);
+  }
+
+  // Transform the raw pointer and return nb::ndarray
+  nb::capsule owner(results, [](void* p) noexcept { delete[] (double*)p; });
+
+  auto result_array =
+      nb::ndarray<double, nb::numpy>(results, output_shape.size(), output_shape.data(), nb::handle(owner)).cast();
+  return result_array;
```

Comment:
This file has grown to a very large entity, maybe its time now to split it into smaller ones ?